### PR TITLE
Initial Client Implementation (DM-2538)

### PR DESCRIPTION
**Pull request by andy-slac**

### CODE REVIEW

File: python/lsst/dax/webserv/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+__author__ = 'bvan'
```
Comment (andy-slac) [inline python/lsst/dax/webserv/__init__.py]: Do we also need `__path__` thing here like in the above files? I guess if we expect other `dax_webserv_*` packages then we do.

File: python/lsst/dax/webserv/client/__init__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+
+__author__ = 'Brian Van Klaveren'
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/__init__.py:21]: Should we for consistency use @author in docstrings everywhere? same for file above.

File: python/lsst/dax/webserv/client/client.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+class MetaClient:
+    """
+    High-level MetaServ Client.
+    See .httpclient.MetaHttpClient for client documentation.
+    """
+
+    def __init__(self, httpClient):
+        self.httpClient = httpClient.meta
+
+    def getRoot(self):
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/client.py]: Docstrings for all public methods would be nice

File: python/lsst/dax/webserv/client/client.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            raise checkedError(e)
+
+
+class Client:
+    """
+    Common endpoint class for all webserv clients
+    """
+    def __init__(self, url, auth_strategy=None, *args, **kwargs):
+        self.httpClient = HttpClient(url, auth_strategy, *args, **kwargs)
+        self.meta = MetaClient(self.httpClient)
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/client.py]: Will this be extended to include clients for two other services?

File: python/lsst/dax/webserv/client/client.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            return resp.json()
+        except RequestException as e:
+            raise checkedError(e)
+
+
+class Client:
+    """
+    Common endpoint class for all webserv clients
+    """
+    def __init__(self, url, auth_strategy=None, *args, **kwargs):
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/client.py]: Need to document all parameters

File: python/lsst/dax/webserv/client/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    def __str__(self):
+        formatted_string = "(HTTP Error: %s): %s" % (self.status_code, self.content)
+        return formatted_string
+
+
+class WebservClientException(WebservException):
+    """
+    The base exception class for all webserv client exceptions.
+    """
+    def __init__(self, exception, message=None, cause=None, metadata=None, **kwargs):
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/exceptions.py:59]: What's kwargs role here, is it just to consume extra parameters in json object?

File: python/lsst/dax/webserv/client/exceptions.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+def checkedError(requestException):
+    """ Check error and return WebservClientException when we can
+    @param requestException: Exception from requests library
+    :return: WebservClientException if it's from the REST API, otherwise WebservRequestException
+    """
+    response = requestException.response
+    if response and response.headers['content-type'] == 'application/json':
+        data = response.json()
+        if 'exception' in data:
+            return WebservClientException(**data)
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/exceptions.py]: This feels a bit unsafe, you depend on data not having some crazy keys. I'd just extract the keys that are needed by WebservClientException and pass them directly.

File: python/lsst/dax/webserv/client/httpclient.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+@author  Brian Van Klaveren, SLAC
+"""
+
+import requests
+from urllib import quote as _qt
+
+class BaseHttpClient:
+    """Request/Response Helper class for all HTTP endpoint clients"""
+
+    def __init__(self, url, auth_strategy=None, *args, **kwargs):
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/httpclient.py]: `*args, **kwargs` are not used, do you need them here?

File: python/lsst/dax/webserv/client/httpclient.py [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            _path = _path.rstrip("/")
+            part = part.lstrip("/")
+            return "/".join((_path, part)) if len(part) else _path
+        newBaseUrl = _resolve(self.base_url, endpoint)
+        return _resolve(newBaseUrl, path)
+
+    def doRequest(self, httpMethod, endpoint, target, params=None, data=None, **kwargs):
+        """If necessary, target should have been quoted"""
+        target = self._target(endpoint, target)
+        headers = kwargs["headers"] if "headers" in kwargs else None
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/httpclient.py:54]: `headers = kwargs.get('headers')`?

File: python/lsst/dax/webserv/client/httpclient.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        requests_method = getattr(requests, httpMethod)
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/httpclient.py:55]: Is httpMethod is expected to be lowercase? This should be documented, usually method name is uppercase. Or force-convert it to lower case here. Or do:

```
resp = requests.request(httpMethod.upper(), ...
```

File: python/lsst/dax/webserv/client/httpclient.py [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+    """
+    MetaServ HTTP Client.
+    This client deals only with requests.Response objects.
+    """
+
+    def __init__(self, baseClient):
+        self.baseClient = baseClient
+        self.endpoint = "/meta"
+
+    def getRoot(self, **kwargs):
```
Comment (andy-slac) [inline python/lsst/dax/webserv/client/httpclient.py]: kwargs?

File: tests/testHttpClient.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+# standard library
+import unittest
+
+from lsst.dax.webserv.client import HttpClient
+
+
+def wrappedMockDoRequest(obj):
+    def mockDoRequest(httpMethod, endpoint, target, params=None, data=None, **kwargs):
+        target = obj._target(endpoint, target)
+        return "{method} {target}".format(method=httpMethod, target=target)
```
Comment (andy-slac) [inline tests/testHttpClient.py]: `(httpMethod + " " + target)` should be somewhat cheaper

File: ups/dax_webserv_client.cfg [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# -*- python -*-
+
+import lsst.sconsUtils
+
+dependencies = {
+    "required": ["python"],
```
Comment (andy-slac) [inline ups/dax_webserv_client.cfg:6]: Do we need anaconda (for requests) as a dependency?

File: ups/dax_webserv_client.table [hunk 14]
```diff
@@ -0,0 +1,5 @@
+setupRequired(requests)
```
Comment (andy-slac) [inline ups/dax_webserv_client.table]: Do we have `requests` as eups package?

File: python/lsst/dax/webserv/client/__init__.py (hunk 2 above)
Comment (brianv0) [inline python/lsst/dax/webserv/client/__init__.py:21]: I'll standardize, my IDE is doing different things (most of this was on old machine, some on new machine)

File: python/lsst/dax/webserv/client/client.py (hunk 4 above)
Comment (brianv0) [inline python/lsst/dax/webserv/client/client.py]: yes, with `self.img` and `self.db` probably.

**Pull request merged**